### PR TITLE
Add xml_missing() when parsing fails

### LINKED ISSUE #1147: Always populate xml_parsed_content and full_xml_parsed_content (perhaps with xml_missing)
Just confirmed that

```
xml2::xml_find_all(xml2::xml_missing, "//a")
# {xml_nodeset (0)}
```

so most of our linters will just work with the replacement.

Unfortunately

```
xml2::xml_find_first(xml2::xml_missing, "//a")
```

is an error, so we may have to guard the code against attempting `xml_find_first()` on missing documents.

### PR DESCRIPTION
**Pull request by MichaelChirico**
Closes #1147

Passing all tests as of now. What other tests would be helpful?

### CODE REVIEW

**Comment by MichaelChirico** [comment]
FYI the meat of the change is in R/is_lint_level.R and R/utils.R

**Comment by MichaelChirico** [comment]
Ran this locally:

```
writeLines("\\", tmp<-tempfile())
lintr::lint(tmp, linters_with_tags(tags=NULL))
l=lintr::lint(tmp, linters_with_tags(tags=NULL))
dput(l)
# structure(list(structure(list(filename = "/tmp/RtmpxgyDZs/file385fe3207fa68a", 
#     line_number = 1L, column_number = 1L, type = "error", message = "unexpected input", 
#     line = "\\", ranges = NULL, linter = NA_character_), class = "lint")), class = "lints")
```

(learned that not _all_ syntax errors lead to `xml_missing()`, but a file with just `\` does)

File: man/default_undesirable_functions.Rd
Comment (MichaelChirico) [inline man/default_undesirable_functions.Rd:56]: missed roxygenize in #1149

**Comment by AshesITR** [comment]
It think that test is all we need.
NB that the current behaviour might be fixed so we should manually zap the XML data from an otherwise good file and expect lint to be silent.

**Comment by MichaelChirico** [comment]
PTAL. It's a bit wonky but I think it'll suffice. I wasn't sure which file to add this test... LMK if you've a better idea.

File: tests/testthat/test-get_source_expressions.R
Comment (AshesITR) [inline tests/testthat/test-get_source_expressions.R:219]: Wouldn't we want to tighten this to `expect_silent()` at least for our own linters?
Also, I can imagine the failure message would be awful to debug. Is this a usecase for {patrick}?

Comment (MichaelChirico) [reply tests/testthat/test-get_source_expressions.R]: `expect_silent()` doesn't quite capture it (since the `a = 1` didn't throw a failure for `assignment_linter()`).

`expect_length(., 0L)` fits the bill better, but because we're using a bit of an awkward interface to lint, we have to use the workaround to #1160 -- `nrow(as.data.frame())`.

File: R/commas_linter.R
Comment (MichaelChirico) [inline R/commas_linter.R:18]: `commas_linter()` was also affected by a similar issue to #1160; the fix here is easy

File: tests/testthat/test-get_source_expressions.R
Comment (AshesITR) [reply tests/testthat/test-get_source_expressions.R]: Oh I think we need to check both, i.e. no output and no lint, then? I'd say we should make our linters consistently return `list()` and the use the simpler test.

Comment (MichaelChirico) [reply tests/testthat/test-get_source_expressions.R]: from typical user perspective it's the same because `flatten_lints()` gives the same output... so i think we're better off leaving that part as a separate fix in #1160.

Comment (MichaelChirico) [reply tests/testthat/test-get_source_expressions.R]: I think we want `lint()` to be silent, i don't think that works with the current test setup.

Comment (AshesITR) [reply tests/testthat/test-get_source_expressions.R]: > from typical user perspective it's the same because `flatten_lints()` gives the same output... so i think we're better off leaving that part as a separate fix in #1160.

I agree it should be a separate PR, just think we should fix it before merging this one so we get a nicer test from the start.

Comment (AshesITR) [reply tests/testthat/test-get_source_expressions.R]: > I think we want `lint()` to be silent, i don't think that works with the current test setup.

I'm just hoping to catch any stray warnings that may or may not be produced by linters when they encounter `xml_missing()` for now.

Comment (MichaelChirico) [reply tests/testthat/test-get_source_expressions.R]: I think we want expect_warning(., NA) then

Comment (MichaelChirico) [reply tests/testthat/test-get_source_expressions.R]: > just think we should fix it before merging this one so we get a nicer test from the start.

I think it's fine, the test is only a little big uglier for now. would agree if it was on the R side.

the main thing is I think the fix will take some effort. don't see a need to hold up this PR (the last code change of the release for now) while we fix internals

Comment (AshesITR) [reply tests/testthat/test-get_source_expressions.R]: Ah thanks for pointing this out. I didn't realize it was that much work.

Is `expect_length(flatten_lints(...), 0L)` an option?

Comment (MichaelChirico) [reply tests/testthat/test-get_source_expressions.R]: flatten_lints not exported right? it'll work but I'm usually hesitant to use unexported stuff. here's probably an ok case as an expedient... is it that much better than nrow(as.df())?

Comment (MichaelChirico) [reply tests/testthat/test-get_source_expressions.R]: > I didn't realize it was that much work.

yea i think it needs a refactor.

current structure is `lapply(foo, function(x) if (bar) then_lint)`. we have to break the inner `if` outside the `lapply` IINM.

Comment (AshesITR) [reply tests/testthat/test-get_source_expressions.R]: It does seem more concise to me, and even more so with a wrapping `expect_warning(..., NA)`...

Fine with either approach as long as we test for warning-free and lint-free results though.

Comment (MichaelChirico) [reply tests/testthat/test-get_source_expressions.R]: OK, I think it should be ready to go

**Review by AshesITR — APPROVED** [review]

**Comment by MichaelChirico** [comment]
Should this get a NEWS item? can add to #833 if so...

**Pull request merged**
